dev/local-environment-dynamic/README.md

```diff
@@ -0,0 +1,135 @@
+
+### Custom Node Image
+
+To use a custom Partner Chains node image, simply update the `PARTNER_CHAINS_NODE_IMAGE` variable at the top of the `setup.sh` script.
+
+## Automated Setup Process Explained
+
+The automation is handled by a series of scripts that execute in a specific order.
+
+### 1. `setup.sh`
+
+This is the main orchestration script. When you run `bash setup.sh`, it performs the following steps:
+
+1.  **System Checks**: Detects the OS and prompts for configuration choices (ports, resource limits, etc.) in interactive mode.
+2.  **Configuration Generation**:
+    -   Creates a `.env` file with all the environment variables for Docker Compose.
+    -   Generates individual entrypoint scripts for each Partner Chains node inside `configurations/partner-chains-nodes/`. These entrypoints include a staggered start delay (`sleep`) to prevent all nodes from starting at once.
+3.  **Docker Compose Manifest Generation**:
+    -   Constructs the main `docker-compose.yml` file by adding service definitions for each Partner Chains node and appending the core Cardano stack (`cardano-node-1`, `postgres`, etc.) from the `modules/` directory.
+
+A key feature of the generated configurations is the use of dynamic discovery flags for the Substrate node:
+
+```bash
+/usr/local/bin/partner-chains-node \\
+  # ... other flags
+  --node-key="$(openssl rand -hex 32)" \\
+  --listen-addr=/ip4/0.0.0.0/tcp/30333 \\
+  --public-addr="/dns4/partner-chains-node-[...]/tcp/30333/p2p/\$PEER_ID" &
+```
+
+- A unique `--node-key` is generated for each node on startup, from which a `PEER_ID` is derived. The node then advertises its public, DNS-based address for discovery.
+
+### 2. `cardano/entrypoint.sh` (inside `cardano-node-1`)
+
+This script bootstraps the Cardano side of the environment.
+
+1.  **Starts Cardano Node**: It waits for configuration files from the setup container, then starts the `cardano-node`.
+2.  **SPO Registration**: It runs a loop that watches the `/shared/spo-certs/` directory. When a new validator's `.cert` file appears, it submits it to the Cardano network, officially registering the validator as an SPO.
+
+### 3. `partner-chains-setup/entrypoint.sh` (inside `partner-chains-setup`)
+
+This is the core script for automating the validator lifecycle.
+
+1.  **Setup**: It waits for the Cardano node and Ogmios, then generates the dynamic Cardano genesis files.
+2.  **Key Generation**: It programmatically generates all necessary keys (payment, stake, cold, VRF, keystore) for every configured validator.
+3.  **Funding**: It builds and submits Cardano transactions to fund the payment addresses of all validators from a genesis UTXO.
+4.  **Certificate Generation**: For each validator, it generates the necessary stake address and stake pool registration certificates, placing the final certificate in `/shared/spo-certs/` to be picked up by the Cardano node.
```

Comment:
Steps 2, 3, and 4 should fully happen only for "registered" validators. Permissioned ones need only polkadot keystores and they don't need anything Cardano related.